// primes-and-factors/build/index.js
"use strict";

var primeFactor = {

  isPrime: function isPrime(num) {
    // Non integer or any number less than 2 is not prime
    if (!Number.isInteger(num) || num < 2) return false;
    // Even number: only prime if it is 2
    if (num % 2 === 0) return num === 2;
    // Odd number divisible by 3: only prime if it is 3
    if (num % 3 === 0) return num === 3;
    // Search for factor 5, 7, 11, 13, 17, 19, 23, 25, 29, 31, 35, 37...
    // up to and including square root of input number
    var floorSqrt = Math.floor(Math.sqrt(num));
    for (var i = 5; i <= floorSqrt; i += 6) {
      if (num % i === 0 || num % (i + 2) === 0) return false;
    }
    return true;
  },

  calculate: function calculate(inputNum) {
    var result = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
    var repeat = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;

    if (!Number.isInteger(inputNum)) return result;
    var num = Math.abs(inputNum);
    if (num < 2) return result;
    var theSqrt = Math.sqrt(num);
    var x = 2;
    if (num % x) {
      x = 3;
      if (num % x) {
        x = 5;
        var add = 2;
        while (num % x && x < theSqrt) {
          // search numbers: 5, 7, 11, 13, 17, 19, 23...
          x += add;
          // add each time: 2, 4, 2, 4, 2, 4, 2...
          add = 6 - add;
        }
      }
    }

    x = x <= theSqrt ? x : num;

    if (!repeat) {
      var index = result.indexOf(x);
      if (index < 0) result.push(x);
    } else result.push(x);

    return x === num ? result : this.calculate(num / x, result, repeat);
  },

  getFactors: function getFactors(num) {
    return this.calculate(num, [], true);
  },

  getUniqueFactors: function getUniqueFactors(num) {
    return this.calculate(num, [], false);
  },

  getPrimeExponentObject: function getPrimeExponentObject(num) {
    var factors = this.calculate(num);
    var countObject = {};
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = factors[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var factor = _step.value;

        if (Number.isFinite(countObject[factor])) {
          countObject[factor] += 1;
        } else {
          countObject[factor] = 1;
        }
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator.return) {
          _iterator.return();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }

    return countObject;
  },

  getFrequency: function getFrequency(num) {
    var countObject = this.getPrimeExponentObject(num);
    var result = [];

    for (var key in countObject) {
      if ({}.hasOwnProperty.call(countObject, key)) {
        var obj = {
          factor: Number(key),
          times: countObject[key]
        };
        result.push(obj);
      }
    }
    return result;
  }

};

module.exports = primeFactor;
